test/e2e/azure.go

```diff
@@ -0,0 +1,223 @@
+}
+
+type azureTestHelper struct {
+	config     *cluserConfig
+	kubeClient client.Client
+	zoneClient dns.ZonesClient
+	zoneName   string
+}
+
+func newAzureHelper(kubeClient client.Client) (providerTestHelper, error) {
+	azureProvider := &azureTestHelper{
+		kubeClient: kubeClient,
+	}
+
+	if err := azureProvider.prepareCredentials(); err != nil {
+		return nil, err
+	}
+
+	if err := azureProvider.prepareZoneClient(); err != nil {
+		return nil, err
+	}
+	return azureProvider, nil
+}
+
+func (a *azureTestHelper) prepareCredentials() (err error) {
+	secret := &corev1.Secret{}
+	secretName := types.NamespacedName{
+		Name:      KUBE_SYSTEM_SECRET_NAME,
+		Namespace: "kube-system",
+	}
+	if err = a.kubeClient.Get(context.Background(), secretName, secret); err != nil {
+		return fmt.Errorf("failed to get credentials secret %s, error : %v", secretName.Name, err)
+	}
+
+	a.config = &cluserConfig{
+		TenantID:                    string(secret.Data[TENANT_ID]),
+		SubscriptionID:              string(secret.Data[SUBSCIPTION_ID]),
+		ResourceGroup:               string(secret.Data[RESOURCE_GROUP]),
+		Location:                    string(secret.Data[REGION]),
+		ClientID:                    string(secret.Data[CLIENT_ID]),
+		ClientSecret:                string(secret.Data[CLIENT_SECRET]),
+		UseManagedIdentityExtension: true,
+	}
+
+	var environment azure.Environment
+	if a.config.Cloud == "" {
+		environment = azure.PublicCloud
+	}
+	a.config.Environment = environment
+
+	return nil
+}
+
+func (a *azureTestHelper) prepareZoneClient() error {
+	token, err := getAccessToken(a.config)
+	if err != nil {
+		return err
+	}
+	a.zoneClient = dns.NewZonesClientWithBaseURI(a.config.Environment.ResourceManagerEndpoint, a.config.SubscriptionID)
+	a.zoneClient.Authorizer = autorest.NewBearerAuthorizer(token)
+	return nil
+}
+
+// Credentials should in as json string for azure provider.
+func (a *azureTestHelper) makeCredentialsSecret(namespace string) *corev1.Secret {
+	credData := struct {
+		TenantID       string `json:"tenantId"`
+		SubscriptionID string `json:"subscriptionId"`
+		ResourceGroup  string `json:"resourceGroup"`
+		ClientID       string `json:"aadClientId"`
+		ClientSecret   string `json:"aadClientSecret"`
+	}{
+		TenantID:       a.config.TenantID,
+		SubscriptionID: a.config.SubscriptionID,
+		ResourceGroup:  a.config.ResourceGroup,
+		ClientID:       a.config.ClientID,
+		ClientSecret:   a.config.ClientSecret,
+	}
+	azureCreds, err := json.Marshal(credData)
+	if err != nil {
+
+	}
+
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("azure-config-file-%s", randomString(16)),
+			Namespace: namespace,
+		},
+		Data: map[string][]byte{
+			"azure.json": azureCreds,
+		},
+	}
+}
+
+func (a *azureTestHelper) platform() string {
+	return string(configv1.AzurePlatformType)
+}
+
+func (a *azureTestHelper) ensureHostedZone(rootDomain string) (string, []string, error) {
```

Comment:
This method is inconsistent with other providers which check for existence of the zone and don't create if it's already there. Not sure it's the best way but let's keep all consistent.

Reply:
yes @alebedev87 , each provider having different SDK,
in aws we have to verify the zone exist or not if exist we have to use else create.
where are in Azure, it will be CreateOrUpdate, SDK iteslf will taken care the duplicates